homeassistant/components/sensor/teleinfo.py

```diff
@@ -0,0 +1,140 @@
+
+DEFAULT_DEVICE = '/dev/ttyUSB0'
+DEFAULT_NAME = 'teleinfo'
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=1)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Optional(CONF_DEVICE, default=DEFAULT_DEVICE): cv.string,
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+TELEINFO_AVAILABLE_VALUES = ['HCHC', 'HCHP', 'IINST', 'IMAX', 'PAPP', 'ISOUSC']
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up for the Teleinfo device."""
+    from kylin import exceptions
+    import kylin
+
+    teleinfo_data = None
+    try:
+        teleinfo = kylin.Kylin(port=config.get(CONF_DEVICE), timeout=2)
+        teleinfo_data = TeleinfoData(teleinfo)
+
+    except exceptions.KylinSerialError as err:
+        _LOGGER.critical("Serial error with Teleinfo: %s", err)
+        return
+
+    if not teleinfo_data.update():
+        _LOGGER.critical("Can't retrieve Teleinfo from device")
+        return
+
+    add_devices([TeleinfoSensor(teleinfo_data, config.get(CONF_NAME))], True)
+
+
+class TeleinfoSensor(Entity):
+    """Implementation of the Teleinfo sensor."""
+
+    def __init__(self, teleinfo_data, name):
+        """Initialize the sensor."""
+        self._name = name
+        self._unit_of_measurement = None
+        self._state = None
+        self._attributes = {
+            ATTR_ATTRIBUTION: ATTRIBUTION,
+        }
+        self._data = teleinfo_data
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the device."""
+        return self._state
+
+    @property
+    def device_state_attributes(self):
+        """Return the state attributes of the device."""
+        return self._attributes
+
+    def update(self):
+        """Get the latest data from Teleinfo device and updates the state."""
+        self._data.update()
+        if self._data.frame:
+            _LOGGER.debug("Frame read: %s", self._data.frame)
+            for info in self._data.frame:
+                if info['name'] in TELEINFO_AVAILABLE_VALUES:
+                    self._attributes[info['name']] = int(info['value'])
+                else:
+                    self._attributes[info['name']] = info['value']
+                if info['name'] == "ADCO":
+                    self._state = self._attributes['ADCO']
+
+
+class TeleinfoData():
+    """Get the latest data from Teleinfo."""
+
+    def __init__(self, teleinfo):
+        """Initialize the data object."""
+        self._frame = None
+        self._teleinfo = teleinfo
+
+    @property
+    def teleinfo(self):
+        """Return the Teleinfo object."""
+        return self._teleinfo
+
+    @property
+    def frame(self):
+        """Return the Teleinfo frame data."""
+        return self._frame
+
+    @Throttle(MIN_TIME_BETWEEN_UPDATES)
+    def update(self):
+        """Get the latest data from Teleinfo device."""
+        self._teleinfo.open()
```

Comment:
Why do we need to open and close the serial connection on each update? I'm thinking we can open the connection in setup_platform and close it when home assistant stops.